File: application/app/gateway/authentication/BasicAuthenticator.scala
package gateway.authentication

import java.nio.charset.StandardCharsets
import java.util.Base64

import domain.model.{Authentication, Endpoint}
import javax.inject.Singleton
import play.api.mvc.RequestHeader
import play.mvc.Http.HeaderNames

import scala.concurrent.Future

@Singleton
class BasicAuthenticator extends Authenticator {
  override def isAuthenticated(request: RequestHeader, endpoint: Endpoint): Future[Boolean] = {
    val auth = endpoint.authentication.asInstanceOf[Authentication.Basic]
    val token = encodeBase64(s"${auth.username}:${auth.password}")
    val expected = s"Basic ${token}"

    request.headers.get(HeaderNames.AUTHORIZATION) match {
      case Some(value) => Future.successful(expected.equals(value))
      case None => Future.successful(false)
    }
  }

  private def encodeBase64(value: String) = {
    Base64.getEncoder.encodeToString(value.getBytes(StandardCharsets.UTF_8))
  }

  override def supports(endpoint: Endpoint): Boolean = endpoint.authentication.isInstanceOf[Authentication.Basic]
}

File: ws-gateway-application/src/main/java/com/cosmin/wsgateway/application/gateway/authentication/BasicAuthenticator.java
package com.cosmin.wsgateway.application.gateway.authentication;

import com.cosmin.wsgateway.domain.Authentication;
import java.nio.charset.StandardCharsets;
import java.util.Base64;
import java.util.Map;
import java.util.Optional;
import org.springframework.stereotype.Component;
import reactor.core.publisher.Mono;


@Component
public class BasicAuthenticator implements Authenticator<Authentication.Basic> {
    @Override
    public Mono<Boolean> isAuthenticated(Map<String, String> headers, Authentication.Basic authentication) {
        return Optional.ofNullable(headers.get("Authorization"))
                .map(rawHeader -> rawHeader.replace("Basic ", ""))
                .map(token -> token.equals(encodeBase64(authentication)))
                .map(Mono::just)
                .orElse(Mono.just(false));
    }

    private String encodeBase64(Authentication.Basic authentication) {
        var value = String.format("%s:%s", authentication.getUsername(), authentication.getPassword());
        return Base64.getEncoder().encodeToString(value.getBytes(StandardCharsets.UTF_8));
    }

    @Override
    public Class<Authentication.Basic> supports() {
        return Authentication.Basic.class;
    }
}
